### PR TITLE
Reuse previous lambda body binding if possible

### PR DESCRIPTION
**Pull request by cston**
Allow re-use of expression-bodied lambda bodies bound during return type inference. (Re-use was disabled in https://github.com/dotnet/roslyn/pull/37052.)

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 1]
```diff
@@ -3195,6 +3193,15 @@ public BoundBlock BindLambdaExpressionAsBlock(ExpressionSyntax body, DiagnosticB
             return bodyBinder.CreateBlockFromExpression(body, bodyBinder.GetDeclaredLocalsForScope(body), refKind, expression, expressionSyntax, diagnostics);
         }
 
+        public BoundBlock BindLambdaExpressionAsBlockContinued(ExpressionSyntax body, BoundExpression expression, DiagnosticBag diagnostics)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:3196]: >BindLambdaExpressionAsBlockContinued [](start = 26, length = 36)

"Continued" doesn't really convey anything specific. Continued from where? Etc. Consider renaming with a more specific name, it feels like simply using ```CreateBlockFromExpression``` would work.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 2]
```diff
@@ -173,19 +170,17 @@ public TypeWithAnnotations GetInferredReturnType(ConversionsBase conversions, Nu
...
                 {
+                    hasReturnValue = true;
                     types.Add((returnStatement.ExpressionOpt, type));
                 }
             }
 
             HashSet<DiagnosticInfo> useSiteDiagnostics = null;
             var bestType = CalculateReturnType(compilation, conversions, delegateType, types, isAsync, node, ref useSiteDiagnostics);
-            return new InferredLambdaReturnType(types.Count, hasReturnWithoutArgument, refKind, bestType, useSiteDiagnostics.AsImmutableOrEmpty());
+            types.Free();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:182]: >types.Free(); [](start = 12, length = 13)

Were we leaking this builder before?

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 3]
```diff
@@ -927,7 +933,7 @@ BoundLambda rebind(BoundLambda lambda)
                     returnType = DelegateReturnTypeWithAnnotations(invokeMethod, out refKind);
                     if (!returnType.HasType || returnType.Type.ContainsTypeParameter())
                     {
-                        var t = (inferredReturnType.HadExpressionlessReturn || inferredReturnType.NumExpressions == 0)
+                        var t = !inferredReturnType.HasReturnValue
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:936]: >!inferredReturnType.HasReturnValue [](start = 32, length = 34)

It feels like this refactoring is not semantically equivalent. It looks like before we could have ```HadExpressionlessReturn == true``` and ```inferredReturnType.NumExpressions > 0``` at the same time. The orignal condition would be ```true```. The new condition for the same scenario would be ```false```.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 4]
```diff
@@ -1216,6 +1222,25 @@ protected override UnboundLambdaState WithCachingCore(bool includeCache)
...
 
+        protected override BoundExpression GetReusableLambdaExpressionBody(BoundBlock body)
+        {
+            if (IsExpressionLambda)
+            {
+                var statements = body.Statements;
+                if (statements.Length == 1 &&
+                    statements[0] is BoundReturnStatement { RefKind: Microsoft.CodeAnalysis.RefKind.None, ExpressionOpt: BoundExpression expr })
+                {
+                    return expr;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:1233]: >return expr; [](start = 20, length = 12)

It is not clear why it is reusable? The name doesn't imply any restrictions on the input. I can imagine a situation when some conversions are have already been applied based on some context, therefore making the expression un-reusable. If the method is not responsible for enforcing reusability, then it should not be part of the name. This is just a helper that extracts return expression from a block with a single ref-none return.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 5]
```diff
@@ -1216,6 +1222,25 @@ protected override UnboundLambdaState WithCachingCore(bool includeCache)
...
+                }
+            }
+            return null;
+        }
+
+        protected override BoundBlock CreateBlockFromExpression(LambdaSymbol lambdaSymbol, Binder lambdaBodyBinder, BoundExpression expression, DiagnosticBag diagnostics)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:1239]: >LambdaSymbol lambdaSymbol [](start = 64, length = 25)

It doesn't look like this parameter aver used, please remove.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 6]
```diff
@@ -541,20 +539,36 @@ private bool DelegateNeedsReturn(MethodSymbol invokeMethod)
...
+            BoundBlock block;
 
-            CSharpCompilation compilation = Binder.Compilation;
-            Binder lambdaBodyBinder = new ExecutableCodeBinder(_unboundLambda.Syntax, lambdaSymbol, ParameterBinder(lambdaSymbol, Binder));
+            var diagnostics = DiagnosticBag.GetInstance();
+            var compilation = Binder.Compilation;
+            var cacheKey = ReturnInferenceCacheKey.Create(delegateType, IsAsync);
+
+            // When binding for real (not for return inference), there is still
+            // a good chance that we could reuse a body of a lambda previous bound for
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:554]: >a good chance that we could reuse a body of a lambda previous bound for [](start = 15, length = 71)

It doesn't look like we are reusing the body, we are reusing some portion of it. Consider reflecting this in the comment and also explaining why it is Ok to reuse that portion.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 7]
```diff
@@ -1216,6 +1222,25 @@ protected override UnboundLambdaState WithCachingCore(bool includeCache)
             return new PlainUnboundLambdaState(unboundLambda: null, Binder, _parameterNames, _parameterIsDiscardOpt, _parameterTypesWithAnnotations, _parameterRefKinds, _isAsync, includeCache);
         }
 
+        protected override BoundExpression GetReusableLambdaExpressionBody(BoundBlock body)
+        {
+            if (IsExpressionLambda)
+            {
+                var statements = body.Statements;
+                if (statements.Length == 1 &&
+                    statements[0] is BoundReturnStatement { RefKind: Microsoft.CodeAnalysis.RefKind.None, ExpressionOpt: BoundExpression expr })
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:1231]: >RefKind: Microsoft.CodeAnalysis.RefKind.None [](start = 60, length = 44)

It is not obvious why this condition is significant, consider adding a clarifying comment.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 8]
```diff
@@ -541,20 +539,36 @@ private bool DelegateNeedsReturn(MethodSymbol invokeMethod)
...
+            var diagnostics = DiagnosticBag.GetInstance();
+            var compilation = Binder.Compilation;
+            var cacheKey = ReturnInferenceCacheKey.Create(delegateType, IsAsync);
+
+            // When binding for real (not for return inference), there is still
+            // a good chance that we could reuse a body of a lambda previous bound for
+            // return type inference.
+            if (_returnInferenceCache.TryGetValue(cacheKey, out BoundLambda returnInferenceLambda) &&
+                GetReusableLambdaExpressionBody(returnInferenceLambda.Body) is BoundExpression expression &&
+                (lambdaSymbol = returnInferenceLambda.Symbol).RefKind == refKind &&
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:558]: >(lambdaSymbol = returnInferenceLambda.Symbol).RefKind == refKind [](start = 16, length = 64)

Should we also check that both are None as ```GetReusableLambdaExpressionBody``` checks below?

File: src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionPerfTests.cs [hunk 9]
```diff
@@ -302,6 +302,43 @@ static void Main()
...
             Enumerable.Range(0, 1).Count(g => true)))))));
     }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+        }
+
+        [ConditionalFactAttribute(typeof(IsRelease))]
+        [WorkItem(1083969, "https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1083969")]
+        public void NestedLambdas_02()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionPerfTests.cs:312]: >NestedLambdas_02 [](start = 20, length = 16)

What exactly are we testing here? Were we unable to compile the code before?

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 10]
```diff
@@ -33,23 +33,20 @@ internal sealed partial class BoundLocalFunctionStatement : IBoundLambdaOrFuncti
...
-            int numExpressions,
-            bool hadExpressionlessReturn,
+            bool hasReturnValue,
             RefKind refKind,
             TypeWithAnnotations typeWithAnnotations,
             ImmutableArray<DiagnosticInfo> useSiteDiagnostics)
         {
-            NumExpressions = numExpressions;
-            HadExpressionlessReturn = hadExpressionlessReturn;
+            HasReturnValue = hasReturnValue;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:49]: >HasReturnValue = hasReturnValue; [](start = 12, length = 32)

It looks like this refactoring is not related to primary goal of this PR, please exclude it from the PR.

File: src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionPerfTests.cs (hunk 9 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionPerfTests.cs:312]: We're testing performance. On my machine, this test completes in a couple of seconds with this change and 15+ mins without.

---
In reply to: [410978595](https://github.com/dotnet/roslyn/pull/43366#discussion_r410978595) [](ancestors = 410978595)

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs (hunk 2 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:182]: Yes.

---
In reply to: [410970357](https://github.com/dotnet/roslyn/pull/43366#discussion_r410970357) [](ancestors = 410970357)

File: src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionPerfTests.cs (hunk 9 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionPerfTests.cs:312]: > We're testing performance. On my machine, this test completes in a couple of seconds with this change and 15+ mins without.

Consider adding a comment to make it easy to analyze failures.

---
In reply to: [411004957](https://github.com/dotnet/roslyn/pull/43366#discussion_r411004957) [](ancestors = 411004957,410978595)

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 11]
```diff
@@ -467,6 +469,17 @@ internal UnboundLambdaState WithCaching(bool includeCache)
         public abstract RefKind RefKind(int index);
         protected abstract BoundBlock BindLambdaBody(LambdaSymbol lambdaSymbol, Binder lambdaBodyBinder, DiagnosticBag diagnostics);
 
+        /// <summary>
+        /// Return the bound expression if the lambda is has an expression body and can be reused easily.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:473]: >is has [](start = 54, length = 6)

Typo?
